src/sknnr_spatial/image/_base.py

```diff
@@ -0,0 +1,193 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Sized
+from functools import singledispatch
+from types import ModuleType
+from typing import TYPE_CHECKING
+
+import numpy as np
+from numpy.typing import NDArray
+
+if TYPE_CHECKING:
+    import xarray as xr
+    from sklearn.base import BaseEstimator
+    from sklearn.neighbors import KNeighborsRegressor
+
+    from ..types import ImageType
+
+
+class ImagePreprocessor(ABC):
+    """
+    A pre-processor for multi-channel image data in a machine learning workflow.
+
+    This class handles flattening an image from 3D pixel space (e.g. y, x, channel)
+    to 2D sample space (sample, channel) to allow prediction and other operations
+    with scikit-learn estimators designed for sample data, and unflattening to
+    convert scikit-learn outputs from sample space back to pixel space.
+
+    Pre-processing can also fill NaN values in sample space to allow prediction with
+    estimators that prohibit NaNs, and mask NoData values in sample space outputs
+    during unflattening to preserve NoData masks from the input image.
```

Comment:
I think we talked through this on Friday, but just for posterity:

- An image may contain "null" pixels represented either by `NaN` or an encoded `NoData` value (e.g. -32768).
- During flattening, only `NaN` values get filled to allow prediction (sklearn estimators don't like `NaNs`).
- After prediction, those "null" pixels get filled with `NaN` in the output. Currently, it's up to the user to replace those with an encoded `NoData` value if they want to use an integer datatype, but that may be a good feature to add in the future since it's likely going to be a common post-processing step.

I think it's going to be worth having a page in the docs devoted to how we handle `NoData`, since it's a tricky topic with a lot of nuances.

Reply:
Yes, that's how I remember our discussion about NoData as well.

> I think it's going to be worth having a page in the docs devoted to how we handle `NoData`, since it's a tricky topic with a lot of nuances.

I fully agree!